src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,391 @@
+                        }
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.STREAMS_AVAILABLE:
+                        {
+                            status = HandleEventStreamsAvailable(
+                                connectionEvent);
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private uint HandleEventConnected(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(_api, _ptr, (uint)QUIC_PARAM_LEVEL.CONNECTION, (uint)QUIC_PARAM_CONN.LOCAL_ADDRESS);
+            _localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(inetAddress);
+
+            _connected = true;
+            _connectTcs.Complete(MsQuicConstants.Success);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownInitiatedByTransport(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (!_connected)
+            {
+                _connectTcs.CompleteException(new IOException("Connection has been shutdown."));
+            }
+
+            _acceptQueue.Writer.Complete();
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownInitiatedByPeer(ConnectionEvent connectionEvent)
+        {
+            _acceptQueue.Writer.Complete();
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            _shutdownTcs.Complete(MsQuicConstants.Success);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventNewStream(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            MsQuicStream msQuicStream = new MsQuicStream(_api, this, connectionEvent.StreamFlags, connectionEvent.Data.NewStream.Stream, inbound: true);
+
+            if (_acceptQueue.Writer.TryWrite(msQuicStream))
+            {
+                if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+                return MsQuicConstants.Success;
+            }
+            else
+            {
+                // Backlog too large, can't accept connections.
+                return MsQuicConstants.InternalError;
```

Comment:
You can't just reject a new stream. You're previously said you're willing to accept X streams at a time. Well, here is one of those stream. If you don't want it, you must go through the correct shutdown process, and provide the corresponding application specific error code to shut it down. This is an absolute MUST.

Reply:
Yeah this was a mistake, streams should always be accepted. I meant to do this only for the listener.